src/Spice86.Core/Emulator/OperatingSystem/DosPathResolver.cs

```diff
@@ -0,0 +1,235 @@
+                return fileNameOnFileSystem;
+            }
+            Regex fileToProcessRegex = FileSpecToRegex(Path.GetFileName(fileToProcess));
+            if (Directory.Exists(parent)) {
+                return Array.Find(Directory.GetFiles(parent), 
+                    x => fileToProcessRegex.IsMatch(x));
+            }
+        } catch (IOException e) {
+            e.Demystify();
+            if (_loggerService.IsEnabled(Serilog.Events.LogEventLevel.Warning)) {
+                _loggerService.Warning(e, "Error while checking file {CaseInsensitivePath}: {Exception}", caseInsensitivePath, e);
+            }
+        }
+
+        return null;
+    }
+    
+    /// <summary>
+    /// Converts a dosVirtualDevices filespec to a regex pattern
+    /// </summary>
+    /// <param name="fileSpec">The DOS filespec</param>
+    /// <returns>The regex pattern</returns>
+    private static Regex FileSpecToRegex(string fileSpec) {
+        string regex = fileSpec.ToLowerInvariant();
+        regex = regex.Replace(".", "[.]");
+        regex = regex.Replace("?", ".");
+        regex = regex.Replace("*", ".*");
+        return new Regex(regex);
+    }
+    
+    /// <inheritdoc />
+    public string? ToHostCaseSensitiveFullName(string dosFileName, bool convertParentOnly) {
+        string fileName = PrefixWithHostDirectory(dosFileName);
+        if (!convertParentOnly) {
+            string? caseSensitivePath = ToCaseSensitiveFileName(fileName, convertParentOnly);
+            return caseSensitivePath;
+        }
+        
+        string? parent = ToCaseSensitiveFileName(fileName, convertParentOnly);
+        if (string.IsNullOrWhiteSpace(parent)) {
+            return null;
+        }
+        
+        return ConvertUtils.ToSlashPath(parent);
+    }
+
+    /// <inheritdoc />
+    public string PrefixWithHostDirectory(string dosPath) {
+        string path = dosPath;
+
+        if(string.IsNullOrWhiteSpace(path)) {
+            return path;
+        }
+
+        if(IsPathRooted(path)) {
+            path = Path.Combine(DriveMap[path[0]].FullName, path[3..]);
+        }
+        else if (StartsWithDosDrive(dosPath)) {
+            path = Path.Combine(CurrentHostDirectory, path[2..]);
+        } else {
+            path = Path.Combine(CurrentHostDirectory, path);
+        }
+
+        return ConvertUtils.ToSlashPath(path);
+    }
+
+    /// <summary>
+    /// All the possible DOS drive letters
+    /// </summary>
+    private static string[] DriveLetters { get; } = { "A", "B", "C", "D", "E", "F", "G", "H", "I", "J", "K", "L", "M", "N", "O", "P", "Q", "R", "S", "T", "U", "V", "W", "X", "Y", "Z" };
+
+    private bool StartsWithDosDrive(string path) => 
+        path.Length >= 2 &&
+        DriveLetters.Contains(path[0].ToString().ToUpperInvariant()) &&
+        path[1] == ':';
+
+    private bool IsPathRooted(string path) => 
+        path.Length >= 3 &&
+        StartsWithDosDrive(path) &&
+        (path[2] == '\\' || path[2] == '/');
```

Comment:
I didn't think DOS paths could have a forward slash?

Also, I don't think `C=\` should be considered "rooted"

Reply:
C:\ is considered by System.IO.Path to be rooted (on Windows) in the .NET BCL.
The same should be true in DOS, as Windows is retro compatible.

This can deal with forward slashes, since the caller can give a parameter that has been already internally converted to forward slashes.

Reply:
I see that Dosbox-x allows both forward and backward slashes. I'm not sure what DOS on real hardware does.

What I meant by `C=\` not being considered rooted, is because `=` is an equals sign, not a colon.
`CC/foo` is also considered "rooted" by this method. I think you should check for a `:` no?

Reply:
That's what StartsWithDosDrive does (checking for ':') and it's called by IsPathRooted.

Can't see how it would consider C=\ to be valid.

Speaking of, we miss DOS invalid characters and invalid names checks.

Reply:
Sorry, misread the dosdrive part. I thouhg it only checked if the drive letter was mounted.

Invalid character and name checks should probably getr a centralized place somewhere once everything is settled a bit.